### DISCUSSION TITLE
Revisit snippet generation tooling for documentation

### DISCUSSION THREAD

**Discussion by maschad** [FuelLabs/fuels-ts#2159] [Ideas]
We experience different inconsistencies with [docs-hub](https://github.com/FuelLabs/docs-hub) - such as https://github.com/FuelLabs/docs-hub/issues/233 and https://github.com/FuelLabs/docs-hub/issues/234 for example - that have created quite a bit of friction. We should perhaps revisit the tooling that we use to generate doc snippets.

The most popular option would probably be [Docusaurus](https://docusaurus.io/)  as it provides a more feature-rich experience out of the box. It has built-in support for versioning, localization, and search, while VitePress requires additional configuration or external tools for these features. Docusaurus also has a larger ecosystem and community support compared to VitePress.

We should at the very least think about localization as we currently only have support for English. Interested to hear others' thoughts

**Comment by arboleya** [comment]
IIRC, we also considered `docsaurus` at the beginning, but since `vitepress` already supports importing regions from text files, I think that's enough for _now_. The scope then extends to code snippets and their sometimes unreliable nature—for example, when using `#context` or `#import` tags. I may have some ideas to draft out and share with you guys.

**Comment by arboleya** [comment]
# Endgoal **TL;DR**

 1. Provide improved separation of concerns:
     1. raw snippets — fully working **_auto-executable_** snippets without regions or cuts
     1. wrapped snippets — wraps raw snippets with HOF at the `post-build` step or similar
     1. tests — usual unit tests, running against the [generated] wrapped snippets
 1. Prohibits the use of code regions
 1. Enforces one text file per code snippet
 1. Snippets are testable and runnable in isolation

# Summary

I'd say the `#context` and `#import` issues are the most pernicious obstacles to correctness and improved DX. We could remove regions from the equation and favor entire files, solving these problems altogether.

I'm also intrigued by running these snippets in isolation.

Could these two ideas converge?

# Hypothesys

Consider a hypothetical snippet that can be:
 1. displayed in the docs
 1. copy + pasted + run locally
 1. click + run online

```ts
// foobar.ts
import { Provider, Wallet, FUEL_NETWORK_URL } from 'fuels';

(async function main () {
  const address = 'fuel..leuf'; // consider the full address here

  const provider = await Provider.create(FUEL_NETWORK_URL);
  const assetId = provider.getBaseAssetId();
  const balance = await Wallet.fromAddress(address, provider).getBalance(assetId);

  console.log(balance);
})();
```


# Test Assertions

How could we ensure that these self-executing functions are working?

 1. Create `.snippets.ts` duplicates wrapped in high-order functions (HOF)
 3. Inject `console.log` capturing hook somehow (just as a )
 4. Write unit tests that target the _generated_ file instead (see below)

Let's say that on `post-build` the generated file would look like this:

> Mind the prepended `export const main = () => ` HOF.

```ts
// foobar.snippet.ts
import { Provider, Wallet, FUEL_NETWORK_URL } from 'fuels';

export const main = () => (async function () {
  const address = 'fuel..leuf'; // consider the full address here

  const provider = await Provider.create(FUEL_NETWORK_URL);
  const assetId = provider.getBaseAssetId();
  const balance = await Wallet.fromAddress(address, provider).getBalance(assetId);

  console.log(balance);
})();
```

Then we could have tests like this asserting the wrapped snippets, not the raw ones:

```ts
// imports wrapped snippet
import { main } from './foobar.snippet.ts';

test('..', () => {
  const res = await main();
  expect(res.logCalls[0]).toEqual(..);
});
```

# Final Snapshot

In the end, we'd have:
 1. `snippets/src/foobar.ts`
 1. `snippets/src/foobar.snisppet.ts` _(generated, ignored)_
 1. `snippets/src/foobar.test.ts`

They'd contain, respectively:
  1. The raw snippet to be shown somewhere (maybe also run?)
  1. The wrapped snippet that tests will use
  1. The tests themselves

# Final Considerations

 1. What problems will removing `#context` and `#import` solve, and how would it improve DX? 
    - It would remove static strings that can go outdated. At the same time, the new model may provide a better DX by offering fully working and auto-executable snippets that one can test, run, and plug into other flows.
 
 1. How would the injection of the `console.log` capturing hook work, and what would be its purpose?
    - It would collect the arguments for every `log` call and return them somehow. Then, unit tests could iterate and assert over them. Please remember that these unit tests would be only concerned with the snippet's output — not their underlying functionality. I'd assume we could even have single-assertion tests for most snippets, which could be easier to maintain.
 
 1. What would be the benefits of having the raw snippet, the wrapped snippet, and the tests in separate files, and how would that improve the project's workflow?
    - This separation of concerns combines all the good parts mentioned above. It gives the ability and confidence to provide helpful snippets that can be used as building blocks, perhaps even without relying on the `see code in context` button, which is currently not implemented in our docs.

**Reply by nedsalk** [reply]
By using ESM modules and top-level await, we can simplify this to one file:
```ts
// foobar.snippet.mts
import { FUEL_NETWORK_URL, Provider, Wallet } from "fuels";

const provider = await Provider.create(FUEL_NETWORK_URL);
const balance = await Wallet.generate({ provider }).getBalances();

console.log(balance);

test("snippet", () => {
  expect(balance.length).toBe(0);
});
```

- Running `pnpm test:filter` on this file correctly evaluates the test (though you'd have to change it to `foobar.snippet.test.mts` currently - this is a modifiable setting),
- When importing the snippets, we remove everything under the first `test(`/`it(` mention and we have the full, self-contained snippet

**Reply by arboleya** [reply]
Separating tests and snippets was intentional.

Top-level awaits may look cleaner, but it also adds constraints.

Pure snippets seem preferable to be displayed and run in isolation.

The idea is to keep special treatment _around_ the snippets' files, not _inside_ them.

**Reply by nedsalk** [reply]
> Separating tests and snippets was intentional.

> The idea is to keep special treatment around the snippets' files, not inside them.

It feels like too much ceremony to separate everything when we can have it all in one file that is verifiable by our testing suite and is parsable by any tool we use to import the snippets into the code. We just have to decide on what is the demarcation between the snippet and the test so that the parsing tool can extract the full snippet. We don't even have to have one if the parsing tool just removes all `test(*)` and `it(*)` blocks - the snippet would be the only thing remaining.
The additional bonus is that we wouldn't need `console.log` when it doesn't make sense to include it in a snippet because the tests have access to the SUT directly - unless you think that every snippet should have a `console.log` regardless.

Based on the premise that we'd put all snippets in the same directory as the markdown file we're importing it into, the folder structure would look like this:
```
- index.md
- getting-balances.snippet.mts
- sending-a-transaction.snippet.mts
```
This is four less files to think about.

> Top-level awaits may look cleaner, but it also adds constraints.

What constraints come to mind? `vitest` readily runs it. We just have to set the project-specific `tsconfig.json`'s compiler settings appropriately (specifically `module` and `target`) so that typescript is happy. I didn't experience any other issues besides this.

> Pure snippets seem preferable to be displayed and run in isolation.

IMO the demarcation between the snippet and test is enough from a maintenance POV. Also note that the tests are just using the results of the snippet above them, they aren't adding any additional preparation steps on their own and would just be a collection of `expect`s upon the snippet.

**Reply by arboleya** [reply]
The demarcation is a trivial thing, and it's not even an issue.

Pure and plain snippets are easier to use, port, and run elsewhere.

They should not be tied with anything (tests, esm custom options, etc).

**Reply by nedsalk** [reply]
I don't fully understand the snippet content of your proposal. How would it be shown to the user on the docs? Wrapped as an IIFE so that it can be copy-pasted and ran right way, or would its content be extracted to show only the body?

**Reply by arboleya** [reply]
Initially, I considered showing it whole, although adding the IFFE brought noise into the party.

The thing about ideals is that once we find a solution to a problem, it usually creates another problem.

**Reply by petertonysmith94** [reply]
> The demarcation is a trivial thing, and it's not even an issue.
> 
> Pure and plain snippets are easier to use, port, and run elsewhere.
> 
> They should not be tied with anything (tests, ESM custom options, etc).

A pure and plain snippet, will also alleviate a lot of the issues that we experience with `docs-hub` not being quite aligned with our documentation as mentioned https://github.com/FuelLabs/docs-hub/issues/233 and https://github.com/FuelLabs/docs-hub/issues/234.

From what I can tell, this would allow the `docs-hub` to be able to use our snippets as is, therefore reducing the need for transforming (which has been a source of our problems currently).

---

I would be nice to not have to generate an intermediary snippet (`snippets/src/foobar.snippet.ts`).

It's worth creating a couple samples solutions, and analysing their pro's and con's.

**Reply by arboleya** [reply]
@petertonysmith94 Here's a simple PoC:
- https://github.com/arboleya/docs-widgets

@nedsalk The constraints are the necessary configurations for using top-level `await`.

**Comment by petertonysmith94** [comment]
I believe this discussion is now closed - we have the following issue:
- https://github.com/FuelLabs/fuels-ts/issues/2504

Please reopen if this asssesment is incorrect 😄
